---
title: "TSSr Vignette"
author: "Zhaolian Lu, Keenan Berry, Zhenbin Hu, Yu Zhan, Tae-Hyuk Ahn, Zhenguo Lin"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{TSSr Vignette}
  %\VignetteEncoding{UTF-8}

---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


Alternatively, you can install the development version directly from GitHub using `devtools`:

```{r github, eval=FALSE}
devtools::install_github("JohnnyChen1113/TSSrcpp")
```

And load `TSSr`:

```{r library, results='hide', message=FALSE}
library(TSSr)
```

# Citation

If you use TSSr, please cite the following article:

```{r citation, eval=TRUE}
citation("TSSr")
```

# Getting help

For general questions about the usage of TSSr, use the [official Bioconductor support forum](https://support.bioconductor.org) and tag your question "TSSr". We strive to answer questions as quickly as possible.

For technical questions, bug reports and suggestions for new features, we refer to the [TSSr github page](https://github.com/JohnnyChen1113/TSSrcpp/issues).

# Quick start

```{r exampleTSSr, results="hide", tidy=FALSE}
# Load the example data
data("exampleTSSr")
myTSSr <- exampleTSSr
```

Or create a new TSSr object

```{r newTSSr, eval=FALSE}
# Provide bam files
inputFiles <- c("S01.sorted.bam", "S02.sorted.bam", "S03.sorted.bam", "S04.sorted.bam")
myTSSr <- new("TSSr", genomeName = "BSgenome.Scerevisiae.UCSC.sacCer3"
          ,inputFiles = inputFiles
          ,inputFilesType= "bam"
          ,sampleLabels = c("SL01","SL02","SL03","SL04")
          ,sampleLabelsMerged = c("control","treat")
          ,mergeIndex = c(1,1,2,2)
          ,refSource = "saccharomyces_cerevisiae.SGD.gff"
          ,organismName = "saccharomyces cerevisiae")
myTSSr
```

```{r getTSS, eval=FALSE}
# Get TSS
getTSS(myTSSr)
```

TSS data processing

```{r TSSprocessing, tidy=FALSE}
# Merge replicates
mergeSamples(myTSSr)
# Normalization
normalizeTSS(myTSSr)
# TSS filtering
filterTSS(myTSSr, method = "TPM", tpmLow = 0.1)
```

TSS clustering

```{r TSSclustering, eval=FALSE}
# TSS clustering 
clusterTSS(myTSSr, method = "peakclu",peakDistance=100,extensionDistance=30
         ,localThreshold = 0.02,clusterThreshold = 1
         ,useMultiCore=FALSE, numCores=NULL)

# Aggregating consensus clusters
consensusCluster(myTSSr, dis = 50, useMultiCore = FALSE)
```

Core promoter shape

```{r shapeCluster, eval=FALSE}
# Calculating core promoter shape score
shapeCluster(myTSSr,clusters = "consensusClusters", method = "PSS",
             useMultiCore= FALSE, numCores = NULL)
```

Annotation core promoters

```{r annotateCluster, tidy=FALSE}
# Assign clusters to the annotated features
annotateCluster(myTSSr,clusters = "consensusClusters",filterCluster = TRUE,
              filterClusterThreshold = 0.02, annotationType = "genes"
              ,upstream=1000, upstreamOverlap = 500, downstream = 0)
```

Differential expression analysis

```{r deGEne, tidy=FALSE}
# Assign clusters to the annotated features
deGene(myTSSr,comparePairs=list(c("control","treat")), 
       pval = 0.01,useMultiCore=FALSE, numCores=NULL)
```

Core promoter shifts

```{r shiftPromoter, tidy=FALSE}
# Calcuate core promoter shifts
shiftPromoter(myTSSr,comparePairs=list(c("control","treat")), pval = 0.01)
```

Session info

```{r sessionInfo, tidy=FALSE}
sessionInfo()
```



